From 22d482a280ade2b421b1e23842e0ccb8b229be24 Mon Sep 17 00:00:00 2001
From: Francois Perrad <francois.perrad@gadz.org>
Date: Sat, 22 May 2010 13:17:03 +0200
Subject: [PATCH] implement setfenv (deprecated with Lua 5.2.0 work3) with debug library

---
 src/Coat.lua |   18 +++++++++++++++++-
 1 files changed, 17 insertions(+), 1 deletions(-)

diff --git a/src/Coat.lua b/src/Coat.lua
index 587cb53..93b80c4 100644
--- a/src/Coat.lua
+++ b/src/Coat.lua
@@ -9,7 +9,6 @@ local pairs = pairs
 local pcall = pcall
 local rawget = rawget
 local require = require
-local setfenv = setfenv
 local setmetatable = setmetatable
 local tostring = tostring
 local basic_type = type
@@ -753,6 +752,23 @@ function with (class, ...)
     end
 end
 
+local function setfenv (level, M)
+    local func = debug.getinfo(level + 1, 'f').func
+    local up = 1
+    while true do
+        local name = debug.getupvalue(func, up)
+        if name == '_ENV' then -- Lua 5.2
+            debug.setupvalue(func, up, M)
+            return
+        end
+        if name == nil then
+            break
+        end
+        up = up + 1
+    end
+    debug.setfenv(func, M) -- Lua 5.1
+end
+
 function module (modname, level)
     if basic_type(package.loaded[modname]) == 'table' then
         error("name conflict for module '" .. modname .. "'")
-- 
1.6.3.3

